source/common/router/router.cc

```diff
@@ -35,7 +35,7 @@ uint32_t getLength(const Buffer::Instance* instance) { return instance ? instanc
 
 void FilterUtility::setUpstreamScheme(Http::HeaderMap& headers,
                                       const Upstream::ClusterInfo& cluster) {
-  if (cluster.sslContext()) {
+  if (cluster.transportSocketFactory().implementsSecureTransport()) {
```

Comment:
Hmmm.  If you have
client->Envoy over HTTP/1.1 and then Envoy1->Envoy2 TLS, I think Envoy1 needs to communicate scheme to Envoy2 (scheme header, or fully qualified URL) and I think Envoy2 needs to not alter the scheme.

I think this will cause problems for you @lizan because we're likely to do HTTP/HTTPS to Envoy1 and non-TLS-crypto to Envoy2 and we'll end up upgrading all the schemes to TLS and losing the original client security level.  I think it's out of scope for this PR but worth thinking about a follow-up for.

Reply:
> Hmmm. If you have
client->Envoy over HTTP/1.1 and then Envoy1->Envoy2 TLS, I think Envoy1 needs to communicate scheme to Envoy2 (scheme header, or fully qualified URL) and I think Envoy2 needs to not alter the scheme.

I could be wrong but AFAIK it's the opposite of this (and in-fact this is what we used to do and it was broken and causing issues). We forward the proto in `x-forwarded-proto` header. `:scheme` needs to be set to the immediate hop (I think).

Reply:
Interesting.  I guess you can do it either way as long as you can consistently tell at each hop is secure and infer what original scheme was :-)

I think then on our side we'll want to move this to an earlier filter (or make it interceptable some other way since you can't edit headers after the router filter) so we can make sure what we send to our upstreams what they're expecting.

Reply:
I would need to get back and look, but I'm fairly positive that the Go gRPC server was refusing requests when using TLS and `:scheme` was not set to HTTPS FYI.

Reply:
@alyssawilk AFAIK `:scheme` is for immediate hop and `x-forwarded-proto` is the original protocol. The original client cert is also forwarded by `x-forwarded-client-cert`.